# CAP 5610 HW3 Task 7
## Robert Forristall

In [1]:
# Import all neccessary libaries
from sklearn.svm import SVC
from sklearn.model_selection import cross_val_score
import numpy as np 
import pandas as pd 

# Import data
training_data = pd.read_csv('train.csv')

# Data Visualization

This section is used to visual the data and gain insight on what features may be important/irrelevant, the type of each feature, and how much missing data is in the training set

In [2]:
print(training_data.info)
training_data.describe()

<bound method DataFrame.info of      PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                   

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

# Preprocessing Data

In [3]:
training_data_X = training_data.drop('Survived', axis='columns')
training_data_Y = training_data['Survived']

training_data_X['Age'] = training_data_X['Age'].fillna(training_data_X['Age'].mean())
training_data_X['Sex'] = training_data_X['Sex'].map({'male': 0, 'female': 1})
training_data_X['Embarked'] = training_data_X['Embarked'].fillna(training_data_X['Embarked'][0])
training_data_X['Embarked'] = training_data_X['Embarked'].map({'S': 0, 'C': 1, 'Q': 2})

# Feature Selection

In [4]:
training_data_X = training_data_X.drop(['PassengerId', 'Name', 'Ticket', 'Cabin', 'Embarked'], axis="columns")

print(training_data_X.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 6 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   Pclass  891 non-null    int64  
 1   Sex     891 non-null    int64  
 2   Age     891 non-null    float64
 3   SibSp   891 non-null    int64  
 4   Parch   891 non-null    int64  
 5   Fare    891 non-null    float64
dtypes: float64(2), int64(4)
memory usage: 41.9 KB
None


# Create SVM model with linear kernel

In [5]:
model_linear = SVC(kernel='linear')
model_linear.fit(training_data_X, training_data_Y)

cross_scores_linear = cross_val_score(model_linear, training_data_X, training_data_Y, cv=5)
print(cross_scores_linear)
print(np.average(cross_scores_linear))

[0.80446927 0.80898876 0.78651685 0.75280899 0.78651685]
0.7878601468834348


# Create SVM model with quadratic kernel

In [6]:
model_poly = SVC(kernel='poly', degree=2)
model_poly.fit(training_data_X, training_data_Y)

cross_scores_poly = cross_val_score(model_poly, training_data_X, training_data_Y, cv=5)
print(cross_scores_poly)
print(np.average(cross_scores_poly))

[0.59776536 0.71348315 0.65730337 0.65730337 0.68539326]
0.6622497018391814


# Create SVM model with RBF kernel

In [7]:
model_rbf = SVC(kernel='rbf')
model_rbf.fit(training_data_X, training_data_Y)

cross_scores_rbf = cross_val_score(model_rbf, training_data_X, training_data_Y, cv=5)
print(cross_scores_rbf)
print(np.average(cross_scores_rbf))

[0.59217877 0.71348315 0.69101124 0.68539326 0.69101124]
0.674615529470843


# Results Analysis

Five Fold Cross Validations
- Linear: [0.80446927 0.80898876 0.78651685 0.75280899 0.78651685]
- Quadratic: [0.59776536 0.71348315 0.65730337 0.65730337 0.68539326]
- RBF: [0.59217877 0.71348315 0.69101124 0.68539326 0.69101124]

Averages of Five Fold Cross Validation
- Linear: 0.7878601468834348
- Quadratic: 0.6622497018391814 
- RBF: 0.674615529470843

Based on the validation scores for my models; it would seem that te best model was the linear model with the RBF being slightly better than quadratic but far behind linear